src/reduction.jl

```diff
@@ -0,0 +1,174 @@
+using LinearAlgebra
+
+""" 
+    canelide(cons_laws::AbstractMatrix, species::AbstractMatrix)
+
+Check if the given species can be simultaneously elided using the given matrix of conservation laws.
+"""
+function canelide(C::AbstractMatrix, species::AbstractVector) 
+    # This amounts to checking that the transformation [ (kept) (elided) ] -> [ (kept) (conserved quantities) ]
+    # is injective, equivalently that the linear map (elided) -> (conserved) is injective,
+    # equivalently that it has full row rank
+    rank(@view C[:,species]) == length(species)  
+end
+
+""" 
+    getinvmap(C::AbstractMatrix, elided_species::AbstractVector)
+
+Return the inverse transformation from observed species and conserved quantities
+to elided species, if it exists (checked via [`canelide`](@ref)), as a pair of matrices
+(M_kept, M_cons). The elided species are recovered via 
+`state_elided = M_kept * state_kept + M_cons * cons`.
+"""
+function getinvmap(C::AbstractMatrix, elided_species::AbstractVector)
+    @assert canelide(C, elided_species)
+    
+    n_elided = length(elided_species)
+    n_kept = size(C, 2) - n_elided
+    n_cons = size(C, 1)
+    kept_species = filter(x -> !(x in elided_species), 1:size(C, 2))
+    M_cons_float = pinv(C[:,elided_species])
+    M_cons = round.(Int, M_cons_float)
```

Comment:
This is really not ideal for numerous reasons. If we want to handle this correctly, we need to use an integer algorithm where no round-off errors occur.

Reply:
I'll see if I can use Smith Normal Form for this, again